File: tools/text_processor.php
<?php
/**
 * Text Processing Tool - PHP Version
 * Provides text manipulation, analysis, and transformation utilities.
 * Compatible with InfinityFree hosting (uses only built-in functions).
 */

class TextProcessor {
    
    /**
     * Count words, characters, and lines in text.
     */
    public static function countWords($text) {
        try {
            $words = preg_split('/\s+/', trim($text));
            $lines = explode("\n", $text);
            $paragraphs = array_filter(preg_split('/\n\n+/', $text), function($p) {
                return trim($p) !== '';
            });
            
            return array(
                "success" => true,
                "word_count" => count($words),
                "character_count" => strlen($text),
                "character_count_no_spaces" => strlen(str_replace(' ', '', $text)),
                "line_count" => count($lines),
                "paragraph_count" => count($paragraphs)
            );
        } catch (Exception $e) {
            return array("success" => false, "error" => $e->getMessage());
        }
    }
    
    /**
     * Find and replace text with optional case sensitivity.
     */
    public static function findAndReplace($text, $find, $replace, $caseSensitive = true) {
        try {
            if ($caseSensitive) {
                $count = substr_count($text, $find);
                $newText = str_replace($find, $replace, $text);
            } else {
                $pattern = '/' . preg_quote($find, '/') . '/i';
                $newText = preg_replace($pattern, $replace, $text);
                $count = preg_match_all($pattern, $text);
            }
            
            return array(
                "success" => true,
                "original_length" => strlen($text),
                "new_length" => strlen($newText),
                "replacements_made" => $count,
                "text" => $newText
            );
        } catch (Exception $e) {
            return array("success" => false, "error" => $e->getMessage());
        }
    }
    
    /**
     * Extract email addresses from text.
     */
    public static function extractEmails($text) {
        try {
            $pattern = '/[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/';
            preg_match_all($pattern, $text, $matches);
            $emails = array_unique($matches[0]);
            
            return array(
                "success" => true,
                "emails" => array_values($emails),
                "count" => count($emails),
                "total_found" => count($matches[0])
            );
        } catch (Exception $e) {
            return array("success" => false, "error" => $e->getMessage());
        }
    }
    
    /**
     * Extract URLs from text.
     */
    public static function extractUrls($text) {
        try {
            $pattern = '/https?:\/\/[^\s<>"{}|\\\\^`\[\]]+/';
            preg_match_all($pattern, $text, $matches);
            $urls = array_unique($matches[0]);
            
            return array(
                "success" => true,
                "urls" => array_values($urls),
                "count" => count($urls),
                "total_found" => count($matches[0])
            );
        } catch (Exception $e) {
            return array("success" => false, "error" => $e->getMessage());
        }
    }
    
    /**
     * Convert text to different cases.
     */
    public static function toCase($text, $caseType) {
        try {
            switch ($caseType) {
                case "upper":
                    $result = strtoupper($text);
                    break;
                case "lower":
                    $result = strtolower($text);
                    break;
                case "title":
                    $result = ucwords(strtolower($text));
                    break;
                case "sentence":
                    $sentences = preg_split('/([.!?]\s*)/', $text, -1, PREG_SPLIT_DELIM_CAPTURE);
                    $result = '';
                    foreach ($sentences as $s) {
                        $result .= ucfirst($s);
                    }
                    break;
                case "camel":
                    $words = preg_split('/\s+/', $text);
                    $result = strtolower($words[0]);
                    for ($i = 1; $i < count($words); $i++) {
                        $result .= ucfirst($words[$i]);
                    }
                    break;
                case "snake":
                    $words = preg_split('/\s+/', $text);
                    $result = strtolower(implode('_', $words));
                    break;
                case "kebab":
                    $words = preg_split('/\s+/', $text);
                    $result = strtolower(implode('-', $words));
                    break;
                default:
                    return array("success" => false, "error" => "Unknown case type: " . $caseType);
            }
            
            return array(
                "success" => true,
                "original" => $text,
                "converted" => $result,
                "case_type" => $caseType
            );
        } catch (Exception $e) {
            return array("success" => false, "error" => $e->getMessage());
        }
    }
    
    /**
     * Generate hash of text.
     */
    public static function hashText($text, $algorithm = "sha256") {
        try {
            $algorithms = array(
                "md5" => "md5",
                "sha1" => "sha1",
                "sha256" => "sha256",
                "sha512" => "sha512"
            );
            
            if (!isset($algorithms[$algorithm])) {
                return array("success" => false, "error" => "Unsupported algorithm: " . $algorithm);
            }
            
            $hash = hash($algorithms[$algorithm], $text);
            
            return array(
                "success" => true,
                "algorithm" => $algorithm,
                "hash" => $hash,
                "input_length" => strlen($text)
            );
        } catch (Exception $e) {
            return array("success" => false, "error" => $e->getMessage());
        }
    }
    
    /**
     * Remove duplicate words or lines.
     */
    public static function removeDuplicates($text, $byLine = false) {
        try {
            if ($byLine) {
                $lines = explode("\n", $text);
                $seen = array();
                $uniqueLines = array();
                foreach ($lines as $line) {
                    if (!in_array($line, $seen)) {
                        $seen[] = $line;
                        $uniqueLines[] = $line;
                    }
                }
                $result = implode("\n", $uniqueLines);
                $removedCount = count($lines) - count($uniqueLines);
            } else {
                $words = preg_split('/\s+/', $text);
                $seen = array();
                $uniqueWords = array();
                foreach ($words as $word) {
                    if (!in_array($word, $seen)) {
                        $seen[] = $word;
                        $uniqueWords[] = $word;
                    }
                }
                $result = implode(' ', $uniqueWords);
                $removedCount = count($words) - count($uniqueWords);
            }
            
            return array(
                "success" => true,
                "original_length" => strlen($text),
                "new_length" => strlen($result),
                "duplicates_removed" => $removedCount,
                "by_line" => $byLine,
                "text" => $result
            );
        } catch (Exception $e) {
            return array("success" => false, "error" => $e->getMessage());
        }
    }
    
    /**
     * Reverse text either character by character or word by word.
     */
    public static function reverseText($text, $byWord = false) {
        try {
            if ($byWord) {
                $words = preg_split('/\s+/', $text);
                $result = implode(' ', array_reverse($words));
            } else {
                $result = strrev($text);
            }
            
            return array(
                "success" => true,
                "original" => $text,
                "reversed" => $result,
                "by_word" => $byWord
            );
        } catch (Exception $e) {
            return array("success" => false, "error" => $e->getMessage());
        }
    }
    
    /**
     * Truncate text to maximum length.
     */
    public static function truncate($text, $maxLength, $suffix = "...") {
        try {
            if (strlen($text) <= $maxLength) {
                return array(
                    "success" => true,
                    "original" => $text,
                    "truncated" => $text,
                    "was_truncated" => false,
                    "original_length" => strlen($text),
                    "max_length" => $maxLength
                );
            }
            
            $truncated = substr($text, 0, $maxLength) . $suffix;
            
            return array(
                "success" => true,
                "original" => $text,
                "truncated" => $truncated,
                "was_truncated" => true,
                "original_length" => strlen($text),
                "truncated_length" => strlen($truncated),
                "max_length" => $maxLength
            );
        } catch (Exception $e) {
            return array("success" => false, "error" => $e->getMessage());
        }
    }
}

// CLI interface for testing
if (php_sapi_name() === 'cli') {
    if ($argc < 2) {
        echo "Usage: php text_processor.php <command> [args...]\n";
        echo "Commands:\n";
        echo "  count <text>              - Count words, chars, lines\n";
        echo "  replace <text> <find> <replace> - Find and replace\n";
        echo "  emails <text>             - Extract emails\n";
        echo "  urls <text>               - Extract URLs\n";
        echo "  case <text> <type>        - Convert case (upper/lower/title/snake/kebab/camel)\n";
        echo "  hash <text> [algorithm]   - Hash text (md5/sha1/sha256/sha512)\n";
        echo "  dedup <text> [--lines]    - Remove duplicates\n";
        echo "  reverse <text> [--words]  - Reverse text\n";
        echo "  truncate <text> <length>  - Truncate text\n";
        exit(1);
    }
    
    $command = $argv[1];
    
    switch ($command) {
        case "count":
            $text = implode(" ", array_slice($argv, 2));
            if (empty($text)) $text = "Sample text for counting";
            $result = TextProcessor::countWords($text);
            break;
        case "replace":
            if ($argc >= 5) {
                $text = $argv[2];
                $find = $argv[3];
                $replace = $argv[4];
            } else {
                $text = "Hello World Hello";
                $find = "Hello";
                $replace = "Hi";
            }
            $result = TextProcessor::findAndReplace($text, $find, $replace);
            break;
        case "emails":
            $text = implode(" ", array_slice($argv, 2));
            if (empty($text)) $text = "Contact us at test@example.com or support@test.org";
            $result = TextProcessor::extractEmails($text);
            break;
        case "urls":
            $text = implode(" ", array_slice($argv, 2));
            if (empty($text)) $text = "Visit https://example.com and http://test.org/page";
            $result = TextProcessor::extractUrls($text);
            break;
        case "case":
            $text = isset($argv[2]) ? $argv[2] : "Hello World";
            $caseType = isset($argv[3]) ? $argv[3] : "upper";
            $result = TextProcessor::toCase($text, $caseType);
            break;
        case "hash":
            $text = isset($argv[2]) ? $argv[2] : "test";
            $algo = isset($argv[3]) ? $argv[3] : "sha256";
            $result = TextProcessor::hashText($text, $algo);
            break;
        case "dedup":
            $text = implode(" ", array_slice($argv, 2));
            if (empty($text)) $text = "hello world hello test world";
            $byLine = in_array("--lines", $argv);
            $result = TextProcessor::removeDuplicates($text, $byLine);
            break;
        case "reverse":
            $text = implode(" ", array_slice($argv, 2));
            if (empty($text)) $text = "Hello World";
            $byWord = in_array("--words", $argv);
            $result = TextProcessor::reverseText($text, $byWord);
            break;
        case "truncate":
            $text = isset($argv[2]) ? $argv[2] : "This is a long text that needs truncation";
            $length = isset($argv[3]) ? intval($argv[3]) : 20;
            $result = TextProcessor::truncate($text, $length);
            break;
        default:
            echo "Unknown command: " . $command . "\n";
            exit(1);
    }
    
    echo json_encode($result, JSON_PRETTY_PRINT) . "\n";
}
?>

File: tools/text_processor.py
#!/usr/bin/env python3
"""
Text Processing Tool - Python Version
Provides text manipulation, analysis, and transformation utilities.
Uses only standard library (no heavy dependencies).
"""

import json
import re
import hashlib
from typing import Dict, List, Any, Optional
from datetime import datetime


class TextProcessor:
    """Text processing and analysis tool."""
    
    @staticmethod
    def count_words(text: str) -> Dict[str, Any]:
        """Count words, characters, and lines in text."""
        try:
            words = text.split()
            lines = text.split('\n')
            
            return {
                "success": True,
                "word_count": len(words),
                "character_count": len(text),
                "character_count_no_spaces": len(text.replace(' ', '')),
                "line_count": len(lines),
                "paragraph_count": len([p for p in text.split('\n\n') if p.strip()])
            }
        except Exception as e:
            return {"success": False, "error": str(e)}
    
    @staticmethod
    def find_and_replace(text: str, find: str, replace: str, case_sensitive: bool = True) -> Dict[str, Any]:
        """Find and replace text with optional case sensitivity."""
        try:
            if case_sensitive:
                new_text = text.replace(find, replace)
                count = text.count(find)
            else:
                pattern = re.compile(re.escape(find), re.IGNORECASE)
                new_text = pattern.sub(replace, text)
                count = len(pattern.findall(text))
            
            return {
                "success": True,
                "original_length": len(text),
                "new_length": len(new_text),
                "replacements_made": count,
                "text": new_text
            }
        except Exception as e:
            return {"success": False, "error": str(e)}
    
    @staticmethod
    def extract_emails(text: str) -> Dict[str, Any]:
        """Extract email addresses from text."""
        try:
            pattern = r'[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}'
            emails = re.findall(pattern, text)
            unique_emails = list(set(emails))
            
            return {
                "success": True,
                "emails": unique_emails,
                "count": len(unique_emails),
                "total_found": len(emails)
            }
        except Exception as e:
            return {"success": False, "error": str(e)}
    
    @staticmethod
    def extract_urls(text: str) -> Dict[str, Any]:
        """Extract URLs from text."""
        try:
            pattern = r'https?://[^\s<>"{}|\\^`\[\]]+'
            urls = re.findall(pattern, text)
            unique_urls = list(set(urls))
            
            return {
                "success": True,
                "urls": unique_urls,
                "count": len(unique_urls),
                "total_found": len(urls)
            }
        except Exception as e:
            return {"success": False, "error": str(e)}
    
    @staticmethod
    def to_case(text: str, case_type: str) -> Dict[str, Any]:
        """Convert text to different cases."""
        try:
            if case_type == "upper":
                result = text.upper()
            elif case_type == "lower":
                result = text.lower()
            elif case_type == "title":
                result = text.title()
            elif case_type == "sentence":
                sentences = re.split(r'([.!?]\s*)', text)
                result = ''.join(s.capitalize() for s in sentences)
            elif case_type == "camel":
                words = re.findall(r'\w+', text)
                result = words[0].lower() + ''.join(word.capitalize() for word in words[1:])
            elif case_type == "snake":
                result = '_'.join(re.findall(r'\w+', text)).lower()
            elif case_type == "kebab":
                result = '-'.join(re.findall(r'\w+', text)).lower()
            else:
                return {"success": False, "error": f"Unknown case type: {case_type}"}
            
            return {
                "success": True,
                "original": text,
                "converted": result,
                "case_type": case_type
            }
        except Exception as e:
            return {"success": False, "error": str(e)}
    
    @staticmethod
    def hash_text(text: str, algorithm: str = "sha256") -> Dict[str, Any]:
        """Generate hash of text."""
        try:
            algorithms = {
                "md5": hashlib.md5,
                "sha1": hashlib.sha1,
                "sha256": hashlib.sha256,
                "sha512": hashlib.sha512
            }
            
            if algorithm not in algorithms:
                return {"success": False, "error": f"Unsupported algorithm: {algorithm}"}
            
            hash_obj = algorithms[algorithm](text.encode('utf-8'))
            
            return {
                "success": True,
                "algorithm": algorithm,
                "hash": hash_obj.hexdigest(),
                "input_length": len(text)
            }
        except Exception as e:
            return {"success": False, "error": str(e)}
    
    @staticmethod
    def remove_duplicates(text: str, by_line: bool = False) -> Dict[str, Any]:
        """Remove duplicate words or lines."""
        try:
            if by_line:
                lines = text.split('\n')
                seen = set()
                unique_lines = []
                for line in lines:
                    if line not in seen:
                        seen.add(line)
                        unique_lines.append(line)
                result = '\n'.join(unique_lines)
                removed_count = len(lines) - len(unique_lines)
            else:
                words = text.split()
                seen = set()
                unique_words = []
                for word in words:
                    if word not in seen:
                        seen.add(word)
                        unique_words.append(word)
                result = ' '.join(unique_words)
                removed_count = len(words) - len(unique_words)
            
            return {
                "success": True,
                "original_length": len(text),
                "new_length": len(result),
                "duplicates_removed": removed_count,
                "by_line": by_line,
                "text": result
            }
        except Exception as e:
            return {"success": False, "error": str(e)}
    
    @staticmethod
    def reverse_text(text: str, by_word: bool = False) -> Dict[str, Any]:
        """Reverse text either character by character or word by word."""
        try:
            if by_word:
                words = text.split()
                result = ' '.join(reversed(words))
            else:
                result = text[::-1]
            
            return {
                "success": True,
                "original": text,
                "reversed": result,
                "by_word": by_word
            }
        except Exception as e:
            return {"success": False, "error": str(e)}
    
    @staticmethod
    def truncate(text: str, max_length: int, suffix: str = "...") -> Dict[str, Any]:
        """Truncate text to maximum length."""
        try:
            if len(text) <= max_length:
                return {
                    "success": True,
                    "original": text,
                    "truncated": text,
                    "was_truncated": False,
                    "original_length": len(text),
                    "max_length": max_length
                }
            
            truncated = text[:max_length] + suffix
            
            return {
                "success": True,
                "original": text,
                "truncated": truncated,
                "was_truncated": True,
                "original_length": len(text),
                "truncated_length": len(truncated),
                "max_length": max_length
            }
        except Exception as e:
            return {"success": False, "error": str(e)}


# CLI interface for testing
if __name__ == "__main__":
    import sys
    
    tp = TextProcessor()
    
    if len(sys.argv) < 2:
        print("Usage: python text_processor.py <command> [args...]")
        print("Commands:")
        print("  count <text>              - Count words, chars, lines")
        print("  replace <text> <find> <replace> - Find and replace")
        print("  emails <text>             - Extract emails")
        print("  urls <text>               - Extract URLs")
        print("  case <text> <type>        - Convert case (upper/lower/title/snake/kebab/camel)")
        print("  hash <text> [algorithm]   - Hash text (md5/sha1/sha256/sha512)")
        print("  dedup <text> [--lines]    - Remove duplicates")
        print("  reverse <text> [--words]  - Reverse text")
        print("  truncate <text> <length>  - Truncate text")
        sys.exit(1)
    
    command = sys.argv[1]
    
    if command == "count":
        text = " ".join(sys.argv[2:]) if len(sys.argv) > 2 else "Sample text for counting"
        result = tp.count_words(text)
    elif command == "replace":
        if len(sys.argv) >= 5:
            text = sys.argv[2]
            find = sys.argv[3]
            replace = sys.argv[4]
        else:
            text = "Hello World Hello"
            find = "Hello"
            replace = "Hi"
        result = tp.find_and_replace(text, find, replace)
    elif command == "emails":
        text = " ".join(sys.argv[2:]) if len(sys.argv) > 2 else "Contact us at test@example.com or support@test.org"
        result = tp.extract_emails(text)
    elif command == "urls":
        text = " ".join(sys.argv[2:]) if len(sys.argv) > 2 else "Visit https://example.com and http://test.org/page"
        result = tp.extract_urls(text)
    elif command == "case":
        text = sys.argv[2] if len(sys.argv) > 2 else "Hello World"
        case_type = sys.argv[3] if len(sys.argv) > 3 else "upper"
        result = tp.to_case(text, case_type)
    elif command == "hash":
        text = sys.argv[2] if len(sys.argv) > 2 else "test"
        algo = sys.argv[3] if len(sys.argv) > 3 else "sha256"
        result = tp.hash_text(text, algo)
    elif command == "dedup":
        text = " ".join(sys.argv[2:]) if len(sys.argv) > 2 else "hello world hello test world"
        by_line = "--lines" in sys.argv
        result = tp.remove_duplicates(text, by_line)
    elif command == "reverse":
        text = " ".join(sys.argv[2:]) if len(sys.argv) > 2 else "Hello World"
        by_word = "--words" in sys.argv
        result = tp.reverse_text(text, by_word)
    elif command == "truncate":
        text = sys.argv[2] if len(sys.argv) > 2 else "This is a long text that needs truncation"
        length = int(sys.argv[3]) if len(sys.argv) > 3 else 20
        result = tp.truncate(text, length)
    else:
        print(f"Unknown command: {command}")
        sys.exit(1)
    
    print(json.dumps(result, indent=2))
